### PR TITLE
BUG: Raise ValueError if a column index in usecols is out of bounds. …

### LINKED ISSUE #25623: ENH/BUG: usecols does not raise an exception when col index is out of bounds.
```python
import pandas as pd

df = pd.DataFrame({'col1': [1, 2], 'col2': [3, 4]})
df.to_csv('test.csv', index=False)

# Should raise ValueError, since there is no col at index 10
pd.read_csv('test.csv', header=0, usecols=[0, 10], engine='python')

print('Silently ran to completion.')
```
<hr>

Output

```python
Silently ran to completion.
```

#### Problem description
Silently passing incorrect column indices (result of some arithmetic) could potentially lead to a debugging nightmare. 

A possible solution is to check if the largest value in `usecols` is larger than the number of columns at any level.

```python
# Something along the lines of
if any(max(col_indices) >= len(column) for column in columns):
    raise ValueError('...')
columns = [[n for i, n in enumerate(column) if i in col_indices]
           for column in columns]
self._col_indices = col_indices
```

If the idea is to accommodate different levels having different number of columns, maybe it would be better if `usecols` allowed a sequence of sequences (one for each level).

**Note**: 
* A similar problem exists with the C engine, although the result is an unexpected column full of _nulls_. 

* The issue seems to be handled correctly when `index` is written to file.

**Edit**:

* The issue occurs when the length of `usecols` equals the number of columns in dataframe.

#### Expected Output
``ValueError``

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.18.0-16-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 4.0.2
pip: 18.1
setuptools: 40.6.3
Cython: 0.29.2
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 7.2.0
sphinx: 1.8.2
patsy: 0.5.1
dateutil: 2.7.5
pytz: 2018.7
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.12
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.2
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.15
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ghost**
fixes #25623

- [x] closes #25623 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -122,7 +122,7 @@ Bug Fixes
 ~~~~~~~~~
 - Bug in :func:`to_datetime` which would raise an (incorrect) ``ValueError`` when called with a date far into the future and the ``format`` argument specified instead of raising ``OutOfBoundsDatetime`` (:issue:`23830`)
 - Bug in an error message in :meth:`DataFrame.plot`. Improved the error message if non-numerics are passed to :meth:`DataFrame.plot` (:issue:`25481`)
--
+- Bug in ``read_csv`` which would not raise ``ValueError`` if a column index in ``usecols`` was out of bounds (:issue:`25623`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.0.rst]: Should we raise an IndexError instead? Though I suppose a ValueError is consistent with the behavior of label `usecols`.

File: doc/source/whatsnew/v0.24.2.rst [hunk 2]
```diff
@@ -18,7 +18,7 @@ including other versions of pandas.
 .. _whatsnew_0242.regressions:
 
 Fixed Regressions
-^^^^^^^^^^^^^^^^^
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.2.rst]: Looks like some git issues. Try reverting the changes to this file.

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (heckeop) [reply doc/source/whatsnew/v0.25.0.rst]: @TomAugspurger `_validate_usecols_names` would raise a `ValueError` if `usecols` contained integer indices and was not equal to the number of columns (`names`) in length. For consistency, it now raises `ValueError` if an index is out bounds when `usecols` has length equal to the number of columns (`names`).

File: doc/source/whatsnew/v0.24.2.rst (hunk 2 above)
Comment (heckeop) [reply doc/source/whatsnew/v0.24.2.rst]: Hmm...The master seems to reflect these changes in v0.24.2.rst as well. Am I missing something?

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.25.0.rst]: I agree with using `ValueError` here.

**Review by gfyoung — APPROVED** [review]
Nice work, @heckeop !

@jreback back over to you

**Comment by jreback** [comment]
lgtm. @heckeop can you merge master. ping on green.

File: pandas/tests/io/parser/test_usecols.py [hunk 3]
```diff
@@ -21,6 +21,20 @@
...
+    (None, [0, 3]),
+    (["a", "b", "c"], [0, -1, 2]),
+    (None, [3]),
+    (["a"], [3])
+])
+def test_usecols_out_of_bounds(all_parsers, names, usecols):
+    data = "a,b,c\n1,2,3\n4,5,6"
+    parser = all_parsers
+
+    with pytest.raises(ValueError, match=_msg_validate_usecols_names):
```
Comment (simonjayhawkins) [inline pandas/tests/io/parser/test_usecols.py]: although this technically doesn't fail the regex, it appears that `_msg_validate_usecols_names` is intended to be used with `.format()`

Comment (WillAyd) [reply pandas/tests/io/parser/test_usecols.py]: Yea good catch @simonjayhawkins

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master. ping on green.

File: pandas/tests/io/parser/test_usecols.py (hunk 3 above)
Comment (heckeop) [reply pandas/tests/io/parser/test_usecols.py]: indeed! @jreback I'll push an update this weekend

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -123,7 +123,7 @@ Bug Fixes
...
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this needs to be moved to the I/O section of bug fixes

**Review by jreback — CHANGES_REQUESTED** [review]
pls merge master and update the whatsnew. ping on green.

**Comment by heckeop** [comment]
@jreback @WillAyd thoughts?

File: doc/source/whatsnew/v0.25.0.rst [hunk 5]
```diff
@@ -182,7 +182,8 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
-
+- Bug in :func:`to_datetime` which would raise an (incorrect) ``ValueError`` when called with a date far into the future and the ``format`` argument specified instead of raising ``OutOfBoundsDatetime`` (:issue:`23830`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: remove these; these have already been moved

File: pandas/tests/io/parser/test_usecols.py [hunk 6]
```diff
@@ -21,6 +21,21 @@
...
 
 
+@pytest.mark.parametrize("names,usecols,missing", [
+    (None, [0, 3], r"\[3\]"),
+    (["a", "b", "c"], [0, -1, 2], r"\[-1\]"),
+    (None, [3], r"\[3\]"),
+    (["a"], [3], r"\[3\]")
+])
+def test_usecols_out_of_bounds(all_parsers, names, usecols, missing):
+    data = "a,b,c\n1,2,3\n4,5,6"
```
Comment (jreback) [inline pandas/tests/io/parser/test_usecols.py]: can you add the issue number here

**Review by jreback — CHANGES_REQUESTED** [review]
doc changes, ping on green.

File: pandas/io/parsers.py [hunk 7]
```diff
@@ -1901,6 +1901,11 @@ def __init__(self, src, **kwds):
                     not set(usecols).issubset(self.orig_names)):
                 _validate_usecols_names(usecols, self.orig_names)
 
+            # GH 25623
```
Comment (jreback) [inline pandas/io/parsers.py]: can you add a 1-line comment here on what you are validating

File: pandas/io/parsers.py [hunk 8]
```diff
@@ -2597,6 +2603,12 @@ def _infer_columns(self):
             if clear_buffer:
                 self._clear_buffer()
 
+            # GH 25623
```
Comment (jreback) [inline pandas/io/parsers.py]: same here

File: pandas/io/parsers.py [hunk 9]
```diff
@@ -2632,6 +2644,10 @@ def _infer_columns(self):
             ncols = len(line)
             num_original_columns = ncols
 
+            # GH25623
```
Comment (jreback) [inline pandas/io/parsers.py]: here as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 7 above)
Comment (heckeop) [reply pandas/io/parsers.py]: @jreback does the comment make sense? I am having a weird writer's block...

Comment (heckeop) [reply pandas/io/parsers.py]: @jreback Added. does the comment make sense?

File: pandas/tests/io/parser/test_usecols.py [hunk 10]
```diff
@@ -21,6 +21,22 @@
...
+    (["a", "b", "c"], [0, -1, 2], r"\[-1\]"),
+    (None, [3], r"\[3\]"),
+    (["a"], [3], r"\[3\]")
+])
+def test_usecols_out_of_bounds(all_parsers, names, usecols, missing):
+    # See gh-25623
+    data = "a,b,c\n1,2,3\n4,5,6"
+    parser = all_parsers
+
+    mssg = _msg_validate_usecols_names.format(missing)
```
Comment (WillAyd) [inline pandas/tests/io/parser/test_usecols.py]: I'm not tied to it but I find the error reporting and naming here somewhat confusing as it doesn't really distinguish position from labels. Is there a logical way to make that distinction here and obviously in the implementation?

Comment (heckeop) [reply pandas/tests/io/parser/test_usecols.py]: I agree, but the purpose of the commit is to provide error reporting consistent with the error raised when the size of _usecols_ is different from the width of the dataframe. 

Distinguishing between reporting mechanisms for labels and indices seems like a new commit.

**Comment by heckeop** [comment]
@jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @heckeop the test is pretty clear i think.

**Pull request merged**